---
title: "Phenoloxidase review"
author: "Leonard Blaschek"
date: "4/1/2021"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ggthemes)
library(showtext)
library(ggbeeswarm)
library(patchwork)
library(broom)
library(ggdendro)
library(dendextend)
library(treeio)
library(ggtree)
library(vegan)
library(phylogram)

#### import Helvetica Neue ####
font_add(
  "Helvetica",
  regular = "/prop_fonts/01. Helvetica   [1957 - Max Miedinger]/HelveticaNeueLTStd-Lt.otf",
  italic = "/prop_fonts/01. Helvetica   [1957 - Max Miedinger]/HelveticaNeueLTStd-LtIt.otf",
  bold = "/prop_fonts/01. Helvetica   [1957 - Max Miedinger]/HelveticaNeueLTStd-Bd.otf",
  bolditalic = "/prop_fonts/01. Helvetica   [1957 - Max Miedinger]/HelveticaNeueLTStd-BdIt.otf"
)
showtext_auto()

#### generating plot theme ####
theme_leo <- function(base_size = 6,
                      base_family = "Helvetica") {
  theme_minimal(
    base_size = base_size,
    base_family = base_family
  ) %+replace%
    theme(
      strip.text = element_text(hjust = 0, face = "italic"),
      axis.ticks = element_line(
        size = 0.125,
        lineend = "square",
        color = "black"
      ),
      axis.text.x = element_text(
        size = 6,
        colour = "black",
        margin = margin(1, 1, 1, 1)
      ),
      axis.text.y = element_text(
        colour = "black",
        size = 6,
        angle = 0,
        vjust = 0.5,
        hjust = 1,
        margin = margin(1, 1, 1, 1)
      ),
      axis.title = element_text(
        colour = "black",
        size = 6
      ),
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      panel.border = element_rect(fill = NA, color = "black", size = 0.25),
      panel.spacing = unit(1.5, "mm"),
      legend.position = "none",
      legend.text = element_text(size = 6),
      legend.key.height = unit(4, "mm"),
      plot.title = element_text(
        size = 6,
        hjust = 0
      ),
      complete = TRUE
    )
}

equal_breaks <- function(n = 3, s = 0.05, ...) {
  function(value) {
    # rescaling
    d <- s * diff(range(value)) / (1 + 2 * s)
    round(seq(0, max(value) - d, length = n), digits = 2)
  }
}

my_breaks <- function(value) {
  if (
    max(value) < 0.005) {
    seq(0, 0.004, length.out = 3)
  } else if (
    max(value) < 0.05) {
    seq(0, 0.02, length.out = 3)
  } else if (
    max(value) < 2) {
    seq(0, 1, length.out = 3)
  } else {
    seq(0, 100, length.out = 3)
  }
}

my_limits <- function(value) {
  if (
    max(value) < 0.005) {
    c(0, 0.0045)
  } else if (
    max(value) < 0.05) {
    c(0, 0.025)
  } else if (
    max(value) < 2) {
    c(-0.25, 1.1)
  } else {
    c(-10, 120)
  }
}

mode <- function(codes) {
  which.max(tabulate(codes))
}

morgenstemning <- c(
  "#000000ff",
  "#051d2eff",
  "#18355fff",
  "#6f2b74ff",
  "#c21b70ff",
  "#f48139ff",
  "#fae400ff",
  "#fefe93ff"
)

pal_ostwald_disc <- c(
  "#275d95",
  "#e8c245",
  "#d25952"
)

ggtext_size <- 6 / (14 / 5)
twocol <- 18 / 2.54
onehalfcol <- 14 / 2.54
onecol <- 9 / 2.54
```

## Evolution and localisation

### Phenoloxidase evolution

```{r}
PO <- read.csv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Data/gene_families.csv") %>%
  select(1:7) %>%
  rename("PRX" = "POX") %>%
  pivot_longer(c("PPO", "LAC", "PRX"), names_to = "variable", values_to = "value") %>%
  mutate(variable = ordered(variable, levels = c("PPO", "LAC", "PRX")))

species_ids <- tibble(
  species_id = c(
    "Cucsa",
    "Manes",
    "Eucgr",
    "Migut",
    "Lus",
    "Prupe",
    "Aqcoe",
    "AL\\dG",
    "evm.model",
    "Glyma",
    "Medtr",
    "LOC_",
    "\\d{5}.m",
    "Sobic",
    "Seita",
    "GSVIV",
    "GRMZM"
  ),
  species = c(
    "Cucumis sativus",
    "Manihot esculenta",
    "Eucalyptus grandis",
    "Mimulus guttatus",
    "Linum usitatissimum",
    "Prunus persica",
    "Aquilegia coerulea",
    "Arabidopsis lyrata",
    "Carica papaya",
    "Glycine max",
    "Medicago truncatula",
    "Oryza sativa",
    "Ricinus communis",
    "Sorghum bicolor",
    "Setaria italica",
    "Vitis vinifera",
    "Zea mays"
  )
)

dirty_blast <- read_tsv(
  "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Data/pblast_gene_fams_tab.out",
  col_names = c("query id", "subject id", "% identity", "alignment length", "mismatches", "gap opens", "q. start", "q. end", "s. start", "s. end", "evalue", "bit score")
) %>%
  mutate(species_id = str_extract(
    `subject id`,
    "Cucsa|Manes|Eucgr|Migut|Lus|Prupe|Aqcoe|AL\\dG|evm.model|Glyma|Medtr|LOC_|\\d{5}.m|Sobic|Seita|GSVIV|GRMZM"
  )) %>%
  left_join(species_ids) %>%
  filter(`bit score` > 300) %>%
  group_by(species) %>%
  summarise(n = n())

annotation <- tribble(
  ~species, ~variable, ~abbrev,
  "Physcomitrella patens", "PRX", "<i>P. patens</i>",
  "Brachypodium distachyon", "PRX", "<i>P. patens</i>"
)

PO_lm <- PO %>%
  group_by(variable) %>%
  filter(genome < 1201) %>%
  do(lin_reg = lm(value ~ genome, data = .)) %>%
  mutate(coefs = list(tidy(lin_reg))) %>%
  unnest(coefs) %>%
  filter(term == "genome")


po_plot <- ggplot(data = PO, aes(x = genome, y = value)) +
  geom_point(aes(size = time, fill = variable),
    colour = "black",
    shape = 21,
    stroke = 0.2,
    alpha = 0.75
  ) +
  scale_size_area(max_size = 5, name = "Divergence [mya]") +
  scale_x_continuous(limits = c(0, 1201)) +
  scale_fill_few(guide = "none") +
  scale_colour_few(name = "Gene family") +
  geom_smooth(aes(colour = variable), method = "lm", se = FALSE, linetype = 2, size = 0.2) +
  guides(
    colour = guide_legend(order = 0),
    size = guide_legend(order = 1)
  ) +
  labs(
    x = "Genome size [Mb]",
    y = "Number of paralogs"
  ) +
  theme_leo() +
  theme(
    legend.position = c(0.25, 0.825),
    legend.box = "horizontal",
    legend.background = element_blank(),
    axis.title.y = element_text(margin = margin(r = -50, "mm"))
  )


pdf("gene_fams.pdf", width = 6, height = 6)
po_plot
dev.off()
```

### Localisation

```{r}
expression <- read.csv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Data/expression.csv") %>%
  mutate(PO = ordered(PO, levels = c("PPO", "LAC", "PRX")))

expression_avg <- expression %>%
  group_by(PO) %>%
  mutate(total = sum(count, na.rm = TRUE)) %>%
  group_by(PO, compartment, method) %>%
  mutate(relative = count / total)

expression_count <- expression %>%
  uncount(weights = count) %>%
  group_by(PO, compartment) %>%
  arrange(method) %>%
  mutate(number = row_number())

ex_plot <- ggplot(expression_count, aes(x = compartment, y = number, fill = PO)) +
  geom_jitter(
    data = filter(expression_count, method == "pred"),
    aes(x = compartment, y = number, fill = PO),
    alpha = 0
  ) +
  annotate("rect",
    ymin = -Inf,
    ymax = Inf,
    xmin = c(0.5, 2.5, 4.5, 6.5),
    xmax = c(1.5, 3.5, 5.5, 7.5),
    fill = "grey95",
    colour = NA
  ) +
  geom_jitter(
    data = filter(expression_count, method == "pred"),
    aes(x = compartment, y = number, fill = PO),
    width = 0.2,
    shape = 21,
    alpha = 0.5,
    size = 0.5,
    stroke = 0.1
  ) +
  geom_jitter(
    data = filter(expression_count, method == "exp"),
    aes(x = compartment, y = number, fill = PO),
    width = 0.2,
    shape = 21,
    alpha = 0.75,
    size = 1,
    stroke = 0.2
  ) +
  labs(y = "Reported protein localisations") +
  theme_leo() +
  theme(axis.title.y = element_blank()) +
  scale_fill_few() +
  facet_wrap(~PO, nrow = 1) +
  coord_flip()

pdf("expression_plot.pdf", height = onecol / 3, width = onecol)
ex_plot
dev.off()
```

### Assemble figure

```{r}
pdf("PO_review_fig2.pdf", height = onecol * 1.1, width = onecol)
po_plot / ex_plot +
  plot_layout(heights = c(3, 1)) &
  plot_annotation(tag_levels = "A") &
  theme(plot.tag = element_text(size = 10, face = "bold", family = "Helvetica"))
```




## Enzyme kinetics

### import data

```{r}
lac_data <- read_csv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Data/lac_data.csv") %>%
  select(1, 2, 8, 10:24) %>%
  pivot_longer(-c("Kingdom", "Species", "Temperature.optimum", "PI"),
    names_to = "variable", values_to = "value"
  ) %>%
  separate(variable, c("var", "substrate"), extra = "merge", remove = TRUE) %>%
  mutate(
    value = as.numeric(value),
    Temperature.optimum = as.numeric(Temperature.optimum),
    Kingdom = ordered(Kingdom, levels = c("Fungi", "Prokaryota", "Plantae"))
  ) %>%
  filter(substrate %in% c("ABTS", "DMP", "SGZ"))
```

### Km figure

```{r}
lac.fig.km <- ggplot(
  data = filter(lac_data, var == "Km"),
  aes(x = Kingdom, y = value, fill = Kingdom)
) +
  geom_quasirandom(
    aes(colour = Kingdom),
    dodge.width = 0.6,
    shape = 16,
    alpha = 0.75,
    size = 1
  ) +
  geom_violin(
    draw_quantiles = 0.5,
    fill = "white",
    colour = "black",
    alpha = 0.5,
    width = 0.4,
    position = position_dodge(width = 0.6),
    size = 0.2,
    scale = "width"
  ) +
  scale_y_log10(limits = c(0.5, 10000)) +
  labs(y = "Km [ÂµM]") +
  # scale_colour_brewer(palette = "Set1") +
  scale_colour_manual(values = pal_ostwald_disc) +
  theme_leo() +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  ) +
  facet_wrap(~substrate)
```

### pH figure

```{r}
lac.fig.ph <- ggplot(data = filter(lac_data, var == "pH"), aes(x = Kingdom, y = value, fill = Kingdom)) +
  geom_quasirandom(
    aes(colour = Kingdom),
    dodge.width = 0.6,
    shape = 16,
    alpha = 0.75,
    size = 1
  ) +
  geom_violin(
    draw_quantiles = 0.5,
    fill = "white",
    colour = "black",
    alpha = 0.5,
    width = 0.4,
    position = position_dodge(width = 0.6),
    size = 0.2,
    scale = "width"
  ) +
  labs(y = "pH optimum") +
  # scale_colour_brewer(palette = "Set1") +
  scale_colour_manual(values = pal_ostwald_disc) +
  scale_y_continuous(limits = c(0, 10), breaks = c(0, 3, 6, 9)) +
  theme_leo() +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  ) +
  facet_wrap(~substrate)
```

### temp figure

```{r}
lac.fig.temp <- ggplot(data = lac_data, aes(x = Kingdom, y = Temperature.optimum, fill = Kingdom)) +
  geom_quasirandom(
    aes(colour = Kingdom),
    dodge.width = 0.6,
    shape = 16,
    alpha = 0.75,
    size = 1
  ) +
  geom_violin(
    draw_quantiles = 0.5,
    fill = "white",
    colour = "black",
    alpha = 0.5,
    width = 0.4,
    position = position_dodge(width = 0.6),
    size = 0.2,
    scale = "width"
  ) +
  annotate("text",
    x = c(1, 2, 3),
    y = c(5, 15, 25),
    label = c("Fungi", "Prokaryota", "Plantae"),
    family = "Helvetica",
    size = ggtext_size
  ) +
  annotate("segment",
    x = c(1, 2, 3),
    xend = c(1, 2, 3),
    y = c(22, 27, 37),
    yend = c(8, 18, 28),
    colour = "black",
    size = 0.2
  ) +
  labs(y = "Temperature optimum [Â°C]") +
  # scale_colour_brewer(palette = "Set1") +
  scale_colour_manual(values = pal_ostwald_disc) +
  scale_y_continuous(limits = c(0, 95)) +
  theme_leo() +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )
```

### pI figure

```{r}
lac.fig.pi <- ggplot(data = lac_data, aes(x = Kingdom, y = PI, fill = Kingdom)) +
  geom_quasirandom(
    aes(colour = Kingdom),
    dodge.width = 0.6,
    shape = 16,
    alpha = 0.75,
    size = 1
  ) +
  geom_violin(
    draw_quantiles = 0.5,
    fill = "white",
    colour = "black",
    alpha = 0.5,
    width = 0.4,
    position = position_dodge(width = 0.6),
    size = 0.2,
    scale = "width"
  ) +
  labs(y = "Isoelectric point") +
  # scale_colour_brewer(palette = "Set1") +
  scale_colour_manual(values = pal_ostwald_disc) +
  scale_y_continuous(limits = c(0, 10.5), breaks = c(0, 3, 6, 9)) +
  theme_leo() +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  )
```

### assemble figure

```{r}
# patch_left <- lac.fig.temp / lac.fig.pi
# patch_right <- lac.fig.km / lac.fig.ph
pdf("lac_fig.pdf", width = onecol, height = onecol * 0.75)
(lac.fig.temp + lac.fig.pi + plot_layout(ncol = 1)) +
  lac.fig.km + lac.fig.ph +
  plot_layout(byrow = FALSE, ncol = 2, widths = c(1, 2)) &
  plot_annotation(tag_levels = "A") &
  theme(plot.tag = element_text(size = 10, face = "bold", family = "Helvetica"))
dev.off()
```

## Homology models

### CASTp

```{r}
# load model quality
model_dope <- read_table("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/2021-03_all_paralogs/model_quality/normalized_DOPE.txt",
  col_names = c("model", "DOPE")
) %>%
  bind_rows(read_table("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/2021-04_described_laccases/model_quality/normalized_DOPE.txt",
    col_names = c("model", "DOPE")
  )) %>%
  bind_rows(read_table("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/AlphaFold2/models/normalized_DOPE.txt",
    col_names = c("model", "DOPE")
  )) %>%
  mutate(model = str_remove(model, fixed(".pdb"))) %>%
  separate(model, into = c("paralog", "model"))

# CASTp job codes
CASTp_codes <- read_csv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/CASTp/job_codes.csv",
  col_names = c("model", "code")
) %>%
  mutate(
    model = str_remove(basename(model), fixed(".pdb")),
    code = str_remove(basename(code), fixed(".zip"))
  ) %>%
  separate(model, into = c("paralog", "model"))

# equivalent residues forming the pocket (manually taken from alignment)
pocket_residues <- read_csv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/pocket_residues.csv") %>%
  pivot_longer(-paralog, names_to = "6klg", values_to = "ID") %>%
  separate(ID, into = c("aa", "position")) %>%
  mutate(position = as.numeric(position),
         position = case_when(paralog == "LcLAC" & position > 323 ~ position - 1,
                              TRUE ~ position))

# load .poc files from CASTp
poc_files <-
  list.files(
    path = "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/CASTp/extracted/",
    pattern = ".poc$",
    recursive = FALSE,
    full.names = TRUE
  )

read_poc <- function(flnm) {
  read_table(flnm,
    col_names = c(
      "type",
      "atom_no",
      "atom",
      "aa",
      "chain",
      "position",
      "x",
      "y",
      "z",
      "who_knows_1",
      "who_knows_2",
      "pocket",
      "pocket_type"
    ),
     col_types = "cncccnnnnnnnc"
  ) %>%
    mutate(
      pocket = case_when(is.na(pocket_type) ~ who_knows_2,
                         TRUE ~ pocket),
      code = str_remove(basename(flnm), fixed(".poc"))
    ) %>%
    select(-who_knows_1, -who_knows_2, -pocket_type)
}

# test <- read_poc("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/CASTp/extracted/j_60f2fb54e839a.poc")

CASTp_poc <- map_dfr(poc_files, read_poc) %>%
  bind_rows() %>%
  left_join(CASTp_codes) %>%
  left_join(pocket_residues)

# identify correct pocket
CASTp_poc_id <- CASTp_poc %>%
  filter(!is.na(`6klg`)) %>%
  group_by(paralog, model, pocket) %>%
  summarise(contact_atoms = n(),
            contact_residues = length(unique(`6klg`))) %>% 
  group_by(paralog, model) %>% 
  filter(contact_residues == max(contact_residues)) %>% # select only the pocket with contact to the maximum number of pocket residues
  filter(contact_atoms == max(contact_atoms)) %>% # if still > 1 pocket, select the one with contact to maximum atoms of pocket residues
  mutate(consensus_pocket = pocket)

# load .pocInfo files from CASTp
pocInfo_files <-
  list.files(
    path = "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/CASTp/extracted/",
    pattern = ".pocInfo",
    recursive = FALSE,
    full.names = TRUE
  )

read_pocInfo <- function(flnm) {
  read_tsv(flnm) %>%
    mutate(
      code = str_remove(basename(flnm), fixed(".pocInfo"))
    )
}

pocInfo <- map(pocInfo_files, read_pocInfo) %>%
  bind_rows()

zmlac3_pocInfo <- read_tsv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/CASTp/ZmLAC3/j_6065d2c555ed5.pocInfo") %>%
  mutate(
    paralog = "ZmLAC3",
    model = "6klg",
    consensus_pocket = 1
  ) %>%
  filter(ID == consensus_pocket)

# load .mouthInfo files from CASTp
mouthInfo_files <-
  list.files(
    path = "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/CASTp/extracted/",
    pattern = ".mouthInfo",
    recursive = FALSE,
    full.names = TRUE
  )

read_mouthInfo <- function(flnm) {
  read_tsv(flnm) %>%
    mutate(
      code = str_remove(basename(flnm), fixed(".mouthInfo"))
    )
}

mouthInfo <- map(mouthInfo_files, read_mouthInfo) %>%
  bind_rows() %>%
  select(Molecule, ID, "mouth_area_sa" = Area_sa)

zmlac3_mouthInfo <- read_tsv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/CASTp/ZmLAC3/j_6065d2c555ed5.mouthInfo") %>%
  mutate(
    paralog = "ZmLAC3",
    model = "6klg",
    consensus_pocket = 1
  ) %>%
  filter(ID == consensus_pocket) %>%
  rename("mouth_area_sa" = Area_sa)

ref_pocket <- zmlac3_pocInfo %>% 
  left_join(select(zmlac3_mouthInfo, paralog, mouth_area_sa), by = "paralog") %>% 
  mutate(
    compactness = ((pi^(1/3)) * ((6 * `Vol_sa`)^(2/3))) / `Area_sa`
  )

pocket_data <- pocInfo %>%
  left_join(mouthInfo) %>%
  left_join(CASTp_codes) %>%
  left_join(CASTp_poc_id) %>%
  left_join(model_dope) %>%
  filter(ID == consensus_pocket) %>%
  filter(!(str_detect(model, "IL"))) %>%  # remove low_quality initial models
  mutate(
    compactness = ((pi^(1/3)) * ((6 * `Vol_sa`)^(2/3))) / `Area_sa`
  )

best_models_castp <- pocket_data %>% 
  group_by(paralog) %>% 
  filter(DOPE == min(DOPE))

write_csv(best_models_castp, "best_models_CASTp.csv")

vol <- ggplot(
  pocket_data,
  aes(
    x = paralog,
    y = Vol_sa
  )
) +
  geom_hline(
    data = zmlac3_pocInfo,
    aes(yintercept = Vol_sa)
  ) +
  geom_quasirandom(
    data = filter(pocket_data, model != "relaxed"),
    aes(colour = DOPE),
    size = 0.5,
    alpha = 0.5) +
  geom_violin(
    draw_quantiles = 0.5,
    alpha = 0.25
  ) +
  geom_point(data = filter(pocket_data, model == "relaxed"),
             aes(fill = DOPE),
             shape = 23,
             stroke = 0.2,
             size = 1.5) +
  scale_colour_gradientn(colours = rev(morgenstemning), limits = c(-2.1, -0.4)) +
  scale_fill_gradientn(colours = rev(morgenstemning), limits = c(-2.1, -0.4), guide = "none") +
  theme_leo() +
  theme(legend.position = "right") +
  coord_flip()

ggplot(
  pocket_data,
  aes(
    x = paralog,
    y = mouth_area_sa
  )
) +
  geom_hline(
    data = zmlac3_mouthInfo,
    aes(yintercept = mouth_area_sa)
  ) +
  geom_quasirandom(aes(colour = DOPE)) +
  geom_violin(
    draw_quantiles = 0.5,
    alpha = 0.25
  ) +
  scale_colour_gradientn(colours = rev(morgenstemning)) +
  theme_leo() +
  theme(legend.position = "right") +
  coord_flip()

pdf("volume_alphafold.pdf", width = 3, height = 4)
vol
dev.off()
```

#### Create .pqr files

```{r}
best_models <- read_csv("/home/leonard/R/PhD/best_alphafold_models.csv")

for(i in 1:nrow(best_models)) {
  model <- best_models[i, "Molecule"]
  pocket <- as.numeric(best_models[i, "pocket"])
  paralog <- best_models[i, "paralog"]
  
  sphere <- jsonlite::fromJSON(paste0("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/CASTp/extracted/", model, ".sphere.json")) # import .sphere.json file
  
  sphere <- tibble(sphere[[pocket]]) %>%  # select correct pocket
    mutate("RecordName" = "ATOM",
           "Serial" = case_when(row_number() < 10 ~ paste0("     ", row_number()),
                                row_number() < 100  ~ paste0("    ", row_number()),
                                TRUE ~ paste0("   ", row_number())),
           "AtomName" = "   O",
           "ResidueName" = "STP",
           # "ChainID" = "A",
           "ResidueNumber" = case_when(pocket < 10 ~ paste0("    ", pocket, "    "),
                                       pocket < 100 ~ paste0("   ", pocket, "    "),
                                       pocket < 1000 ~ paste0("  ", pocket, "    "),
                                       TRUE ~ paste0("  ", pocket, "    ")),
           "X" = format(round(center$x, 3), width = 7),
           "Y" = format(round(center$y, 3), width = 7),
           "Z" = format(round(center$z, 3), width = 7),
           "Charge" = "0.00",
           "Radius" = round(radius, 2)) %>% 
    select(-c(1, 2)) %>% 
    add_row("RecordName" = "TER") %>% 
    add_row("RecordName" = "END")
  
  write_delim(sphere, 
              paste0(paralog, ".pqr"), 
              col_names = FALSE, na = "", delim = " ", quote = "none")
}
```



### fpocket

```{r}
# function to read .txt files
read_info <- function(flnm) {
  read_tsv(flnm,
    col_names = c("pocket", "variable", "value"),
    col_types = "ccn",
    trim_ws = TRUE
  ) %>%
    fill(pocket) %>%
    drop_na() %>%
    mutate(
      file = str_remove(basename(flnm), fixed("_info.txt")),
      variable = str_remove(variable, "[:space:]*:"),
      pocket = as.numeric(str_extract(pocket, "[:digit:]+"))
    ) %>%
    separate(file, into = c("paralog", "model"))
}

# function to read .atm files
read_atm <- function(flnm) {
  read_csv(flnm,
    comment = "HEADER",
    col_names = c(
      "type", "atmn_no", "atm_type",
      "aa", "chain", "position",
      "x", "y", "z", "occupancy",
      "temp", "element", "just_a_zero"
    )
  ) %>%
    filter(type == "ATOM") %>% 
    mutate(file = flnm)
}


# load model quality
model_dope <- read_table("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/2021-03_all_paralogs/model_quality/normalized_DOPE.txt",
  col_names = c("model", "DOPE")
) %>%
  bind_rows(read_table("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/2021-04_described_laccases/model_quality/normalized_DOPE.txt",
    col_names = c("model", "DOPE")
  )) %>%
  mutate(model = str_remove(model, fixed(".pdb"))) %>%
  separate(model, into = c("paralog", "model"))

# equivalent residues forming the pocket (manually taken from alignment)
pocket_residues <- read_csv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/pocket_residues.csv") %>%
  select(-c(2:4)) %>% 
  pivot_longer(-paralog, names_to = "6klg", values_to = "ID") %>%
  separate(ID, into = c("aa", "position")) %>%
  mutate(position = as.numeric(position))

# paths to fpocket output
path1 <- "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/2021-04_described_laccases/"
path2 <- "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/2021-03_all_paralogs/"

# # load _info.txt files from fpocket
# info_files <-
#   list.files(
#     path = c(
#       path1,
#       path2
#     ),
#     pattern = ".txt$",
#     recursive = TRUE,
#     full.names = TRUE
#   )
# 
# pocket_info <- map(info_files, read_info) %>%
#   bind_rows()
# 
# write_csv(pocket_info, "pocket_info.csv")

pocket_info_defaults <- read_csv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/pocket_info_defaults.csv") %>% 
  mutate(run = "defaults")

pocket_info_D2 <- read_csv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/pocket_info_D2.csv") %>% 
  mutate(run = "D2")

pocket_info_D1_6 <- read_csv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/pocket_info.csv") %>% 
  mutate(run = "D1.6")

pocket_info <- bind_rows(pocket_info_defaults, pocket_info_D2, pocket_info_D1_6)

# # load _atm.csv files made from the fpocket _atm.pdb files
# atm_files <-
#   list.files(
#     path = c(
#       path1,
#       path2
#     ),
#     pattern = "_atm.csv$",
#     recursive = TRUE,
#     full.names = TRUE
#   )
# 
# pocket_atm <- map(atm_files, read_atm) %>%
#   data.table::rbindlist()
# 
# write_csv(
#   pocket_atm,
#   "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/fpocket_D2_results.csv"
# )

pocket_atm_1_6 <- read_csv(
  "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/fpocket_results.csv"
) %>% 
  mutate(run = "D1.6")

pocket_atm_2 <- read_csv(
  "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/fpocket_D2_results.csv"
) %>% 
  mutate(run = "D2")

pocket_atm_defaults <- read_csv(
  "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/fpocket_defaults_results.csv"
) %>% 
  mutate(run = "defaults")

pocket_atm <- bind_rows(
  pocket_atm_1_6, 
  pocket_atm_2, 
  pocket_atm_defaults
  )

pocket <- pocket_atm %>%
  drop_na() %>%
  mutate(
    file = str_remove(file, paste0(path1, "|", path2)),
    file = str_remove(file, "ZmLAC3/")
  ) %>%
  separate(file, into = c("blank", "model", "folder", "pocket"), sep = "/") %>%
  separate(model, into = c("paralog", "model"), sep = "\\.") %>%
  mutate(
    model = str_remove(model, fixed("_out")),
    pocket = as.numeric(str_extract(pocket, "[:digit:]+"))
  ) %>%
  select(-blank, -folder) %>%
  left_join(pocket_residues) %>%
  drop_na() %>%
  filter(!(str_detect(model, "IL"))) %>% # remove low_quality initial models
  unite("model", model, run) %>% 
  group_by(paralog, model, pocket) %>%
  summarise(contact_atoms = n(),
            contact_residues = length(unique(`6klg`))) %>% 
  filter(contact_residues > 5) %>% 
  separate(model, into = c("model", "run"), sep = "_") %>% 
  group_by(paralog, model) %>% 
  filter(contact_residues == max(contact_residues)) %>% # select only the pocket with contact to the maximum number of pocket residues
  filter(contact_atoms == max(contact_atoms)) %>% # if still > 1 pocket, select the one with contact to maximum atoms of pocket residues
  left_join(pocket_info) %>%
  group_by(paralog, model) %>% 
  pivot_wider(names_from = variable, values_from = value) %>% 
  filter(Volume == min(Volume)) %>%  # if still > 1 pocket, select smallest pocket
  mutate(
    compactness = ((pi^(1/3)) * ((6 * `Volume`)^(2/3))) / `Total SASA`
  )

ref_pocket <- pocket %>%
  filter(paralog == "ZmLAC3")

plotly::ggplotly(ggplot(
  pocket,
  aes(
    x = reorder(paralog, -Volume),
    y = Volume
  )
) +
  geom_quasirandom(aes(colour = model)) +
  geom_violin(
    draw_quantiles = 0.5,
    alpha = 0.25
  ) +
  # scale_colour_gradientn(colours = rev(morgenstemning)) +
  theme_leo() +
  theme(
    legend.position = "right",
    plot.margin = unit(c(5, 1, 1, 1), "mm")
  ))
# geom_curve(data = zmlac3_depth,
#            aes(xend = 22.75,
#                yend = `mean(depth)`,
#                x = 23.5,
#                y = `mean(depth)` + 1),
#            curvature = 0.4,
#            size = 0.2,
#            arrow = arrow(length = unit(0.01, "npc"))) +
# geom_text(data = zmlac3_depth,
#            aes(x = 23.5,
#                y = `mean(depth)` + 1),
#           label = "ZmLAC3",
#           hjust = 0,
#           family = "Helvetica",
#           size = ggtext_size) +
# coord_flip(clip = "off", xlim = c(1, 22))

```

#### CASTp vs fpocket

```{r}
comp <- pocket_data %>% 
  select(paralog, model, Vol_sa) %>% 
  left_join(pocket)

ggplot(comp,
       aes(x = Volume,
           y = Vol_sa)) +
  geom_point() +
  geom_smooth() +
  coord_fixed()
```


### DEPTH

```{r}
# T1 coordinating histidines (manually taken from alignment)
coord_his <- read_csv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/pocket_residues.csv") %>%
  pivot_longer(-paralog, names_to = "6klg", values_to = "ID") %>%
  filter(`6klg` %in% c("HIS_451", "HIS_519")) %>%
  separate(ID, into = c("aa", "position")) %>%
  mutate(position = as.numeric(position),
         # because I removed an X from the LcLAC sequence for alphafold
         position = case_when(paralog == "LcLAC" & position > 323 ~ position - 1,
                              TRUE ~ position)) 

# load atom.depth files from DEPTH
depth_files <-
  list.files(
    path = c(
      "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/2021-03_all_paralogs/DEPTH/",
      "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/2021-04_described_laccases/DEPTH/",
      "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/AlphaFold2/DEPTH/"
      ),
    pattern = "atom.depth",
    recursive = FALSE,
    full.names = TRUE
  )

read_depth <- function(flnm) {
  read_tsv(flnm) %>%
    mutate(
      model = str_remove(basename(flnm), fixed(".pdb-atom.depth"))
    ) %>%
    separate(model, into = c("paralog", "model"))
}

zmlac3_depth <- read_depth("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/2021-03_all_paralogs/DEPTH/ZmLAC3/6klg.pdb-atom.depth") %>%
  mutate(paralog = "ZmLAC3",
         model = "6klg")

depth <- map(depth_files, read_depth) %>%
  bind_rows() %>%
  bind_rows(zmlac3_depth) %>%
  left_join(coord_his) %>%
  left_join(model_dope)

his_depth <- depth %>%
  separate(residue, into = c("chain", "residue")) %>%
  filter(!(str_detect(model, "IL"))) %>% # remove low_quality initial models
  filter(residue == position & `atom-type` == "NE2")

write_csv(his_depth, "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/his_depths.csv")

his_depth <- read_csv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/his_depths.csv")

zmlac3_depth <- his_depth %>%
  filter(paralog == "ZmLAC3")

his_depth <- his_depth %>%
  filter(paralog != "ZmLAC3")

ggplot(
  his_depth,
  aes(
    x = reorder(paralog, -`mean(depth)`),
    y = `mean(depth)`
  )
) +
  geom_hline(
    data = zmlac3_depth,
    aes(yintercept = `mean(depth)`)
  ) +
  geom_quasirandom(aes(colour = DOPE)) +
  geom_violin(
    draw_quantiles = 0.5,
    alpha = 0.25
  ) +
  scale_colour_gradientn(colours = rev(morgenstemning)) +
  theme_leo() +
  theme(
    legend.position = "right",
    plot.margin = unit(c(5, 1, 1, 1), "mm")
  ) +
  facet_wrap(~`6klg`,
    ncol = 2
  ) +
  geom_curve(
    data = zmlac3_depth,
    aes(
      xend = 22.75,
      yend = `mean(depth)`,
      x = 23.5,
      y = `mean(depth)` + 1
    ),
    curvature = 0.4,
    size = 0.2,
    arrow = arrow(length = unit(0.01, "npc"))
  ) +
  geom_text(
    data = zmlac3_depth,
    aes(
      x = 23.5,
      y = `mean(depth)` + 1
    ),
    label = "ZmLAC3",
    hjust = 0,
    family = "Helvetica",
    size = ggtext_size
  ) +
  coord_flip(clip = "off", xlim = c(1, 22))
```

### Combine data (fpocket)

```{r}
# model_data <- his_depth %>%
#   select(paralog, model, `6klg`, `mean(depth)`, DOPE) %>%
#   pivot_wider(names_from = `6klg`, values_from = `mean(depth)`) %>%
#   left_join(pocket %>%
#     select(
#       paralog,
#       model,
#       Volume,
#       compactness,
#       `Charge score`,
#       `Hydrophobicity score`,
#       `Polarity score`
#     )) %>%
#   pivot_longer(c(
#     HIS_451,
#     HIS_519,
#     Volume,
#     compactness,
#     `Charge score`,
#     `Hydrophobicity score`,
#     `Polarity score`
#   ),
#   names_to = "parameter", values_to = "value"
#   ) %>%
#   mutate(paralog = recode(paralog, "LcLAC" = "ADE/LAC"))
# 
# corr_data <- model_data %>%
#   pivot_wider(names_from = parameter, values_from = value) %>%
#   drop_na() %>%
#   select(-paralog, -model, -DOPE) %>%
#   cor() %>%
#   corrplot::corrplot()
# 
# ref_data <- zmlac3_depth %>%
#   select(paralog, model, `6klg`, `mean(depth)`) %>%
#   pivot_wider(names_from = `6klg`, values_from = `mean(depth)`) %>%
#   left_join(ref_pocket %>%
#     select(
#       paralog,
#       model,
#       Volume,
#       compactness,
#       `Charge score`,
#       `Hydrophobicity score`,
#       `Polarity score`
#     )) %>%
#   pivot_longer(c(
#     HIS_451,
#     HIS_519,
#     Volume,
#     compactness,
#     `Charge score`,
#     `Hydrophobicity score`,
#     `Polarity score`
#   ),
#   names_to = "parameter", values_to = "value"
#   )
# 
# model_data <- model_data %>%
#   mutate(
#     parameter = ordered(parameter, levels = c(
#       "HIS_451",
#       "HIS_519",
#       "Volume",
#       "compactness",
#       "Charge score",
#       "Polarity score",
#       "Hydrophobicity score"
#     ))
#   )
# 
# ref_data <- ref_data %>%
#   mutate(parameter = ordered(parameter, levels = c(
#     "HIS_451",
#     "HIS_519",
#     "Volume",
#     "compactness",
#     "Charge score",
#     "Polarity score",
#     "Hydrophobicity score"
#   )))
# 
# best_models <- model_data %>% 
#   mutate(paralog = recode(paralog, "ADE/LAC" = "LcLAC")) %>% 
#   group_by(paralog) %>% 
#   filter(DOPE == max(DOPE)) %>% 
#   left_join(pocket) %>% 
#   filter(parameter == "compactness")
# 
# write_csv(best_models, "best_fpocket_models.csv")
```


### Combine data (CASTp)

```{r}
model_data <- his_depth %>%
  select(paralog, model, `6klg`, `mean(depth)`, DOPE) %>%
  pivot_wider(names_from = `6klg`, values_from = `mean(depth)`) %>%
  left_join(pocket_data %>%
    select(
      paralog,
      model,
      Vol_sa,
      compactness,
      mouth_area_sa
    )) %>%
  pivot_longer(c(
    HIS_451,
    HIS_519,
    Vol_sa,
    compactness,
    mouth_area_sa
  ),
  names_to = "parameter", values_to = "value"
  ) %>%
  mutate(paralog = recode(paralog, "LcLAC" = "ADE/LAC")) %>%
  filter(model == "relaxed")

corr_data <- model_data %>%
  pivot_wider(names_from = parameter, values_from = value) %>%
  drop_na() %>%
  select(-paralog, -model, -DOPE) %>%
  cor() %>%
  corrplot::corrplot()

ref_data <- zmlac3_depth %>%
  select(paralog, model, `6klg`, `mean(depth)`) %>%
  pivot_wider(names_from = `6klg`, values_from = `mean(depth)`) %>%
  left_join(ref_pocket %>%
    select(
      paralog,
      model,
      Vol_sa,
      compactness,
      mouth_area_sa
    )) %>%
  pivot_longer(c(
    HIS_451,
    HIS_519,
    Vol_sa,
    compactness,
    mouth_area_sa
  ),
  names_to = "parameter", values_to = "value"
  )

model_data <- model_data %>%
  mutate(
    parameter = ordered(parameter, levels = c(
      "Vol_sa",
      "compactness",
      "mouth_area_sa",
      "HIS_451",
      "HIS_519"
    ))
  ) %>%
  mutate(kingdom = "plantae")

ref_data <- ref_data %>%
  mutate(parameter = ordered(parameter, levels = c(
    "Vol_sa",
    "compactness",
    "mouth_area_sa",
    "HIS_451",
    "HIS_519"
  )))

best_models <- pocket_data %>% 
  group_by(paralog) %>% 
  filter(DOPE == min(DOPE))

write_csv(best_models, "best_alphafold_models.csv")
```

### Assemble figure

```{r}
clust_data <- model_data %>%
  group_by(paralog, parameter) %>%
  summarise(value = median(value, na.rm = T)) %>%
  pivot_wider(names_from = parameter, values_from = value) %>%
  column_to_rownames(var = "paralog")

clust <- reorder(
  hclust(dist(clust_data), method = "average"),
  -clust_data$Vol_sa,
  agglo.FUN = "mean"
)

gg_clust <- dendro_data(as.dendrogram(clust))

model_data <- model_data %>%
  mutate(paralog = ordered(paralog, levels = gg_clust[["labels"]][["label"]]))

dendrogram <- ggplot(gg_clust$segments) +
  geom_segment(aes(x = x, y = y, xend = xend, yend = yend),
    lineend = "square",
    size = 0.2
  ) +
  scale_y_reverse() +
  scale_x_continuous(expand = c(0.026, 0.025)) +
  coord_flip() +
  theme_void()

axis_titles <- c(
  Vol_sa = "Volume [Ã<sup>3</sup>]",
  compactness = "Compactness",
  mouth_area_sa = "Mouth area [Ã<sup>2</sup>]",
  HIS_451 = "Depth of His 451 [Ã]",
  HIS_519 = "Depth of His 519 [Ã]"
)

facet_limits <- tibble(
  parameter = ordered(
    c(
      "Vol_sa",
      "compactness",
      "mouth_area_sa",
      "HIS_451",
      "HIS_519"
    ),
    levels = c(
      "Vol_sa",
      "compactness",
      "mouth_area_sa",
      "HIS_451",
      "HIS_519"
    )
  ),
  limits = c(
    300,
    0.7,
    34,
    20,
    14
  )
)

violins <- ggplot(
  model_data,
  aes(
    x = paralog,
    y = value
  )
) +
  geom_hline(
    data = ref_data,
    aes(yintercept = value),
    size = 0.2,
    linetype = 2
  ) +
  geom_point(aes(fill = DOPE, group = model),
    shape = 21,
    alpha = 1,
    size = 2,
    stroke = 0.2
  ) +
  geom_point(
    data = tibble(
      "paralog" = "ChLAC8",
      "parameter" = ordered(
        "mouth_area_sa",
        levels = c(
          "Vol_sa",
          "compactness",
          "mouth_area_sa",
          "HIS_451",
          "HIS_519"
        )
      ),
      "value" = 70.924
    ),
    alpha = 0
  ) +
  # geom_violin(
  #   draw_quantiles = 0.5,
  #   alpha = 0.25,
  #   colour = "black",
  #   size = 0.2
  # ) +
  geom_curve(
    data = ref_data %>% filter(parameter == "Vol_sa"),
    aes(
      xend = 22.75,
      yend = value + 0.1,
      x = 23.5,
      y = value + 20
    ),
    curvature = 0.4,
    size = 0.2,
    arrow = arrow(length = unit(0.03, "npc"))
  ) +
  geom_text(
    data = ref_data %>% filter(parameter == "Vol_sa"),
    aes(
      x = 23.5,
      y = value + 25
    ),
    label = "ZmLAC3",
    hjust = 0,
    family = "Helvetica",
    size = ggtext_size
  ) +
  geom_text(
    data = tibble(
      parameter =
        ordered(c("HIS_519", "HIS_519"),
          levels = c(
            "Vol_sa",
            "compactness",
            "mouth_area_sa",
            "HIS_451",
            "HIS_519"
          )
        )
    ),
    x = 25,
    y = c(0, 13),
    label = c("better", "worse"),
    hjust = 1,
    family = "Helvetica",
    size = ggtext_size
  ) +
  # geom_segment(
  #   data = tibble(parameter = ordered(c("compactness", "mouth_area_sa"),
  #     levels = c(
  #       "HIS_451",
  #       "HIS_519",
  #       "Vol_sa",
  #       "compactness",
  #       "mouth_area_sa"
  #     )
  #   )),
  #   aes(
  #     xend = 24,
  #     yend = c(0.7, 25),
  #     x = 24,
  #     y = c(0.9, 15)
  #   ),
  #   size = 0.2,
  #   arrow = arrow(length = unit(0.03, "npc"))
  # ) +
  scale_fill_gradientn(colours = rev(morgenstemning), limits = c(-2.1, -1.2)) +
  expand_limits(y = 0) +
  labs(fill = "Model quality\n(DOPE Z-score)") +
  theme_leo() +
  theme(
    legend.position = c(0.83, 1.05),
    # legend.position = "none",
    legend.direction = "horizontal",
    legend.key.height = unit(2, "mm"),
    legend.title = element_text(vjust = 1),
    plot.margin = unit(c(5.5, 1, 0, 1), "mm"),
    axis.title = element_blank(),
    strip.placement = "outside",
    axis.ticks.x = element_blank(),
    axis.text.x = element_blank(),
    strip.text = element_blank()
    # strip.text = ggtext::element_markdown(
    #   hjust = 0.5, face = "plain",
    #   margin = unit(c(1, 1, 1, 1), "mm")
    # )
  ) +
  coord_flip(clip = "off", xlim = c(1, 22)) +
  facet_wrap(~parameter,
    nrow = 1,
    scales = "free_x",
    strip.position = "bottom",
    labeller = as_labeller(axis_titles)
  )

# plotly::ggplotly(violins)

pdf("homology_model_fig.pdf", height = onecol, width = twocol)
dendrogram + violins +
  plot_layout(widths = c(1, 5))
dev.off()
```

### Fungal laccases

#### CASTp

```{r}

# load model quality
model_dope <- read_table(
  "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/Fungal/models/normalized_DOPE.txt",
  col_names = c("model", "DOPE")
  ) %>% 
  mutate(model = str_remove(model, fixed(".pdb")))

# CASTp job codes
CASTp_codes <- read_csv(
  "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/Fungal/CASTp/job_codes.csv",
  col_names = c("model", "code")
  ) %>%
  mutate(
    model = str_remove(basename(model), fixed(".pdb")),
    code = str_remove(basename(code), fixed(".zip"))
  )

# equivalent residues forming the pocket (manually taken from alignment)
pocket_residues <- read_csv(
  "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/Fungal/pocket_residues.csv"
  ) %>% 
  mutate(comment = "pocket residue") %>% 
  # select(aa, comment)
  filter(aa %in% c("HIS", "ASP", "LEU"))

# load .poc files from CASTp
poc_files <-
  list.files(
    path = "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/Fungal/CASTp/",
    pattern = ".poc$",
    recursive = FALSE,
    full.names = TRUE
  )

read_poc <- function(flnm) {
  read_table(flnm,
    col_names = c(
      "type",
      "atom_no",
      "atom",
      "aa",
      "chain",
      "position",
      "x",
      "y",
      "z",
      "who_knows_1",
      "who_knows_2",
      "pocket",
      "pocket_type"
    ),
     col_types = "cncccnnnnnnnc"
  ) %>%
    mutate(
      pocket = case_when(is.na(pocket_type) ~ who_knows_2,
                         TRUE ~ pocket),
      code = str_remove(basename(flnm), fixed(".poc"))
    ) %>%
    select(-who_knows_1, -who_knows_2, -pocket_type)
}

# test <- read_poc("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/CASTp/extracted/j_60f2fb54e839a.poc")

CASTp_poc <- map_dfr(poc_files, read_poc) %>%
  left_join(CASTp_codes) %>%
  left_join(pocket_residues)

# identify correct pocket
CASTp_poc_id <- CASTp_poc %>%
  group_by(model, pocket) %>%
  summarise(contact_atoms = n(),
            contact_residues = length(unique(comment))) %>% 
  group_by(model) %>% 
  filter(contact_residues == max(contact_residues)) %>% # select only the pocket with contact to the maximum number of pocket residues
  filter(contact_atoms == max(contact_atoms)) %>% # if still > 1 pocket, select the one with contact to maximum atoms of pocket residues
  mutate(consensus_pocket = pocket)

# load .pocInfo files from CASTp
pocInfo_files <-
  list.files(
    path = "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/Fungal/CASTp/",
    pattern = ".pocInfo",
    recursive = FALSE,
    full.names = TRUE
  )

read_pocInfo <- function(flnm) {
  read_tsv(flnm) %>%
    mutate(
      code = str_remove(basename(flnm), fixed(".pocInfo"))
    )
}

pocInfo <- map_dfr(pocInfo_files, read_pocInfo)

# load .mouthInfo files from CASTp
mouthInfo_files <-
  list.files(
    path = "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/Fungal/CASTp/",
    pattern = ".mouthInfo",
    recursive = FALSE,
    full.names = TRUE
  )

read_mouthInfo <- function(flnm) {
  read_tsv(flnm) %>%
    mutate(
      code = str_remove(basename(flnm), fixed(".mouthInfo"))
    )
}

mouthInfo <- map_dfr(mouthInfo_files, read_mouthInfo) %>%
  select(Molecule, ID, "mouth_area_sa" = Area_sa)

pocket_data_fungus <- pocInfo %>%
  left_join(mouthInfo) %>%
  left_join(CASTp_codes) %>%
  left_join(CASTp_poc_id) %>%
  left_join(model_dope) %>%
  filter(ID == consensus_pocket) %>%
  mutate(
    compactness = ((pi^(1/3)) * ((6 * `Vol_sa`)^(2/3))) / `Area_sa`
  ) %>% 
  drop_na()

vol <- ggplot(
  pocket_data_fungus,
  aes(
    x = model,
    y = Vol_sa
  )
) +
  geom_point(aes(fill = DOPE),
             shape = 23,
             stroke = 0.2,
             size = 1.5) +
  scale_colour_gradientn(colours = rev(morgenstemning), limits = c(-2.1, -0.4)) +
  scale_fill_gradientn(colours = rev(morgenstemning), limits = c(-2.1, -0.4), guide = "none") +
  theme_leo() +
  theme(legend.position = "right") +
  coord_flip()

mouth <- ggplot(
  pocket_data_fungus,
  aes(
    x = model,
    y = mouth_area_sa
  )
) +
  geom_point(aes(fill = DOPE),
             shape = 23,
             stroke = 0.2,
             size = 1.5) +
  scale_colour_gradientn(colours = rev(morgenstemning), limits = c(-2.1, -0.4)) +
  scale_fill_gradientn(colours = rev(morgenstemning), limits = c(-2.1, -0.4), guide = "none") +
  theme_leo() +
  theme(legend.position = "right") +
  coord_flip()

pdf("volume_alphafold.pdf", width = 6, height = 4)
vol + mouth
dev.off()

```

#### Create .pqr files

```{r}
best_models <- pocket_data

for(i in 1:nrow(best_models)) {
  model <- best_models[i, "code"]
  pocket <- as.numeric(best_models[i, "pocket"])
  paralog <- best_models[i, "model"]
  
  # import .sphere.json file
  sphere <- jsonlite::fromJSON(
    paste0(
      "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/Fungal/CASTp/",
      model,
      ".sphere.json"
      )
    )
  
  sphere <- tibble(sphere[[pocket]]) %>%  # select correct pocket
    mutate("RecordName" = "ATOM",
           "Serial" = case_when(row_number() < 10 ~ paste0("     ", row_number()),
                                row_number() < 100  ~ paste0("    ", row_number()),
                                TRUE ~ paste0("   ", row_number())),
           "AtomName" = "   O",
           "ResidueName" = "STP",
           # "ChainID" = "A",
           "ResidueNumber" = case_when(pocket < 10 ~ paste0("    ", pocket, "    "),
                                       pocket < 100 ~ paste0("   ", pocket, "    "),
                                       pocket < 1000 ~ paste0("  ", pocket, "    "),
                                       TRUE ~ paste0("  ", pocket, "    ")),
           "X" = format(round(center$x, 3), width = 7),
           "Y" = format(round(center$y, 3), width = 7),
           "Z" = format(round(center$z, 3), width = 7),
           "Charge" = "0.00",
           "Radius" = round(radius, 2)) %>% 
    select(-c(1, 2)) %>% 
    add_row("RecordName" = "TER") %>% 
    add_row("RecordName" = "END")
  
  write_delim(sphere, 
              paste0(paralog, ".pqr"), 
              col_names = FALSE, na = "", delim = " ", quote = "none")
}
```

#### DEPTH

```{r}
# load atom.depth files from DEPTH
depth_files <-
  list.files(
    path = "/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Modelling/Fungal/DEPTH/",
    pattern = "atom.depth",
    recursive = FALSE,
    full.names = TRUE
  )

read_depth <- function(flnm) {
  read_tsv(flnm) %>%
    mutate(
      model = str_remove(basename(flnm), fixed(".pdb-atom.depth"))
    )
}

depth <- map_dfr(depth_files, read_depth) %>%
  left_join(model_dope) %>% 
  filter(!str_detect(model, "unrelaxed"))

his_depth_fungus <- depth %>%
  separate(residue, into = c("chain", "residue")) %>%
  filter(residue %in% c(458, 395) & `atom-type` == "NE2")

ggplot(
  his_depth_fungus,
  aes(
    x = reorder(model, -`mean(depth)`),
    y = `mean(depth)`
  )
) +
  geom_point() +
  scale_colour_gradientn(colours = rev(morgenstemning)) +
  theme_leo() +
  theme(
    legend.position = "right",
    plot.margin = unit(c(5, 1, 1, 1), "mm")
  ) +
  facet_wrap(~residue,
    ncol = 2
  ) +
  coord_flip(clip = "off")
```

#### figure

```{r}
model_data_fungus <- his_depth_fungus %>%
  select(model, residue, `mean(depth)`, DOPE) %>%
  pivot_wider(names_from = residue, values_from = `mean(depth)`) %>%
  rename("HIS_451" = `395`,
         "HIS_519" = `458`) %>% 
  left_join(pocket_data_fungus %>%
    select(
      model,
      Vol_sa,
      compactness,
      mouth_area_sa
    )) %>%
  pivot_longer(c(
    HIS_451,
    HIS_519,
    Vol_sa,
    compactness,
    mouth_area_sa
  ),
  names_to = "parameter", values_to = "value"
  ) %>% 
  mutate(
    parameter = ordered(parameter, levels = c(
      "Vol_sa",
      "compactness",
      "mouth_area_sa",
      "HIS_451",
      "HIS_519"
    )),
    paralog = recode(
      str_remove(model, fixed("_relaxed_model_1")),
      "1kya_1" = "1KYA (X-ray)",
      "1KYA_AP2" = "1KYA (AF2)",
      "F162A_F332A" = "F162A/F332A"),
    model = "relaxed",
    kingdom = "fungi"
  )

clust_data <- model_data_fungus %>%
  group_by(paralog, parameter) %>%
  summarise(value = median(value, na.rm = T)) %>%
  pivot_wider(names_from = parameter, values_from = value) %>%
  column_to_rownames(var = "paralog")

clust <- reorder(
  hclust(dist(clust_data), method = "average"),
  -clust_data$Vol_sa,
  agglo.FUN = "mean"
)

gg_clust <- dendro_data(as.dendrogram(clust))

model_data_fungus <- model_data_fungus %>%
  mutate(paralog = ordered(paralog, levels = gg_clust[["labels"]][["label"]]))

dendrogram_fungi <- ggplot(gg_clust$segments) +
  geom_segment(aes(x = x, y = y, xend = xend, yend = yend),
    lineend = "square",
    size = 0.2
  ) +
  scale_y_reverse() +
  scale_x_continuous(expand = c(0.09, 0.09)) +
  coord_flip() +
  theme_void()

axis_titles <- c(
  Vol_sa = "Volume [Ã<sup>3</sup>]",
  compactness = "Compactness",
  mouth_area_sa = "Mouth area [Ã<sup>2</sup>]",
  HIS_451 = "Depth of His 451 [Ã]",
  HIS_519 = "Depth of His 519 [Ã]"
)

violins_fungi <- ggplot(
  model_data_fungus,
  aes(
    x = paralog,
    y = value
  )
) +
  geom_hline(
    data = ref_data,
    aes(yintercept = value),
    size = 0.2,
    linetype = 2
  ) +
  geom_point(aes(fill = DOPE, group = model),
    shape = 21,
    alpha = 1,
    size = 2,
    stroke = 0.2
  ) +
  geom_point(
    data = tibble(
      paralog = "1KYA (X-ray)",
      parameter = ordered(c(
        "Vol_sa",
        "compactness",
        "mouth_area_sa",
        "HIS_451",
        "HIS_519"
      ),
      levels = c(
        "Vol_sa",
        "compactness",
        "mouth_area_sa",
        "HIS_451",
        "HIS_519"
      )
      ),
      value = c(
        max(model_data$value[model_data$parameter == "Vol_sa"]),
        max(model_data$value[model_data$parameter == "compactness"]),
        max(model_data$value[model_data$parameter == "mouth_area_sa"]),
        max(model_data$value[model_data$parameter == "HIS_451"]),
        max(model_data$value[model_data$parameter == "HIS_519"])
      )
    ),
    alpha = 0
  ) +
  scale_fill_gradientn(colours = rev(morgenstemning), limits = c(-2.1, -1.2)) +
  expand_limits(y = 0) +
  labs(fill = "Model quality\n(DOPE Z-score)") +
  theme_leo() +
  theme(
    # legend.position = c(0.83, 1.05),
    legend.position = "none",
    legend.direction = "horizontal",
    legend.key.height = unit(2, "mm"),
    legend.title = element_text(vjust = 1),
    plot.margin = unit(c(0, 1, 1, 1), "mm"),
    axis.title = element_blank(),
    strip.placement = "outside",
    strip.text = ggtext::element_markdown(
      hjust = 0.5, face = "plain",
      margin = unit(c(1, 1, 1, 1), "mm")
    )
  ) +
  coord_flip() +
  facet_wrap(~parameter,
    nrow = 1,
    scales = "free_x",
    strip.position = "bottom",
    labeller = as_labeller(axis_titles)
  )

# plotly::ggplotly(violins)

pdf("homology_model_fig_fungus.pdf", height = onecol * 0.5, width = twocol)
dendrogram_fungi + violins_fungi +
  plot_layout(widths = c(1, 5))
dev.off()
```

### Combined figure

```{r}
layout <- "
ABBBBB
ABBBBB
ABBBBB
CDDDDD
"

pdf("homology_model_fig_complete.pdf", height = onecol * 0.8, width = twocol)
dendrogram + violins + dendrogram_fungi + violins_fungi +
  plot_layout(design = layout)
dev.off()

```


## Phylogeny

### read tree files

```{r}

# tree of modelled paralogs
model_tree <- read.nexus("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Phylogeny/2021-04-18_modelled_paralogs/output/rooted.nxs")

# tree of blasted paralogs
phylogeny <- read.mrbayes("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Phylogeny/2020-04-14_mrbayes/flat_prior/removed_X/no_Beagle/rooted.nxs")

phylogeny_labels <- read_delim("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Phylogeny/2020-04-14_mrbayes/flat_prior/removed_X/no_Beagle/short_names.txt",
  delim = " ",
  col_names = c("taxon", "ID")
) %>%
  mutate(taxon = str_extract(taxon, "[:digit:]+"))

phylogeny_species <- read_delim("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Phylogeny/2020-04-14_mrbayes/flat_prior/removed_X/no_Beagle/species.txt",
  col_names = c("type", "taxon", "species"),
  delim = "="
) %>%
  mutate(
    taxon = str_extract(taxon, "[:digit:]+"),
    paralog = str_extract(species, "[:digit:]+"),
    species = case_when(
      str_detect(species, fixed("high_confidence")) ~ "[Picea abies]",
      str_detect(species, fixed("Azolla")) ~ "[Azolla filiculoides]",
      str_detect(species, fixed("Zea")) ~ "[Zea mays]",
      str_detect(species, fixed("Arabidopsis")) ~ "[Arabidopsis thaliana]",
      TRUE ~ species
    ),
    species = str_extract(species, "\\[.+\\]"),
    species = str_extract(species, "[:alpha:]+ [:alpha:]+"),
    short_species = paste0(
      str_extract(species, "^[:alpha:]"),
      str_extract(species, " [:alpha:]") %>% str_remove(" ")
    ),
    paralog = paste0(short_species, "LAC", paralog),
    paralog = case_when(
      str_detect(paralog, "ZmLAC") ~ "ZmLAC3",
      str_detect(species, "Chlamy") ~ "Chlamyd.",
      TRUE ~ str_replace(paralog, "LACNA", "")
    ),
    species_colour = case_when(
      species %in% c(
        "Zea mays",
        "Chlamydomonas reinhardtii",
        "Miscanthus sinensis"
      ) ~ NA_character_,
      TRUE ~ species
    )
  ) %>%
  left_join(phylogeny_labels) %>%
  select(ID, species, species_colour, paralog)

phylo_n <- phylogeny_species %>%
  group_by(species) %>%
  summarise(n = n())

# phylogeny_bak <- phylogeny

# phylogeny <- as.treedata(ape::root(phylogeny, node = 407, edgelabel = TRUE))
# phylogeny@data <- phylogeny_bak@data
```

### Phylogeny figure

```{r}
spacer <- 6.75
text_spacer <- 1

phylo_plot <- ggtree(phylogeny, branch.length = "none", layout = "circular") %<+%
  phylogeny_species +
  aes(colour = species_colour) +
  geom_tippoint(aes(
    subset = (!species %in% c("Zea mays", "Miscanthus sinensis", "Chlamydomonas reinhardtii")),
    fill = factor(species)
  ),
  shape = 21,
  colour = "black",
  size = 1.5,
  stroke = 0.2
  ) +
  scale_fill_brewer(palette = "Spectral") +
  scale_colour_brewer(
    palette = "Spectral",
    na.value = "black"
  ) +
  # geom_tiplab(aes(subset = (label %in% c("tr_Q2PAJ1_Q2PAJ1_MAIZE-49-584",
  #                                        "QED40958.1-45-583",
  #                                        "XP_001690061.1-319-821"))),
  #                 label = c("ZmLAC3",
  #                           "MsLAC1",
  #                           "CrLAC1"))
  geom_tiplab2(aes(
    subset = (species %in% c(
      "Arabidopsis thaliana",
      "Zea mays",
      "Chlamydomonas reinhardtii",
      "Miscanthus sinensis"
    )),
    label = paralog
  ),
  hjust = -0.1,
  family = "Helvetica",
  fontface = "italic",
  size = ggtext_size
  ) +
  geom_cladelabel(
    node = 416,
    label = "Ascorbate oxidases",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  # geom_cladelabel(node = 231,
  #                 label = "Basal laccases",
  #                 color = "black",
  #                 offset = 5,
  #                 offset.text = text_spacer,
  #                 align = TRUE,
  #                 angle = "auto",
  #                 horizontal = FALSE,
  #                 hjust = 0.5,
  #                 fontsize = ggtext_size,
  #                 fontface = "bold",
  #                 family = "Helvetica"
  #                 ) +
  geom_cladelabel(
    node = 242,
    label = "I",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  geom_cladelabel(
    node = 348,
    label = "II",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  geom_cladelabel(
    node = 394,
    label = "III",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  geom_cladelabel(
    node = 363,
    label = "IV",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  geom_cladelabel(
    node = 322,
    label = "V",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  geom_cladelabel(
    node = 302,
    label = "VI",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  geom_cladelabel(
    node = 260,
    label = "VII",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  # geom_label2(aes(subset = (prob > 0.9 & !isTip),
  #                label = round(as.numeric(prob), digits =2)),
  #            size = 2,
  #            family = "Helvetica",
  #            label.size = NA,
  #            label.padding = unit(0, "mm"),
  #            nudge_x = -1,
  #            nudge_y = 1) +
  # geom_text(aes(label = node), size = 2) +
  # geom_point2(aes(subset = (node == 1)), colour = "black", size = 5) +
  xlim(-10, NA) +
  guides(
    colour = "none",
    # fill = "none"
    fill = guide_legend(label.position = "left")
  ) +
  theme(
    text = element_text(family = "Helvetica"),
    legend.text = element_text(face = "italic", size = 6),
    legend.title = element_blank(),
    legend.position = c(0.85, 0.1),
    legend.text.align = 1,
    legend.background = element_rect(fill = NA),
    legend.key.size = unit(2, "mm"),
    plot.margin = unit(c(0, 0, 0, 0), "mm")
  )

pdf("lac_phylogeny.pdf", width = onecol * 1.5, height = onecol * 1.5)
phylo_plot
dev.off()
```

#### Phylogeny supplement
```{r}

#detailed tree
spacer <- 0.75
text_spacer <- 0.1

phylo_plot <- ggtree(phylogeny, layout = "circular") %<+%
  (phylogeny_species %>% mutate("ID2" = ID)) +
  aes(colour = species_colour) +
  geom_treescale(x = 0.3, y = 225, offset = -7, family = "Helvetica") +
  geom_tippoint(aes(
    subset = (!species %in% c("Zea mays", "Miscanthus sinensis", "Chlamydomonas reinhardtii")),
    fill = factor(species)
  ),
  shape = 21,
  colour = "black",
  size = 1.5,
  stroke = 0.2
  ) +
  scale_fill_brewer(palette = "Spectral") +
  scale_colour_brewer(
    palette = "Spectral",
    na.value = "black"
  ) +
  geom_tiplab2(
    aes(label = paste0(" ", ID2, "â", species, " ")),
    colour = "black",
    hjust = 0,
    family = "Helvetica",
    fontface = "italic",
    size = ggtext_size
  ) +
  geom_cladelabel(
    node = 416,
    label = "Ascorbate oxidases",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  # geom_cladelabel(node = 231,
  #                 label = "Basal laccases",
  #                 color = "black",
  #                 offset = 5,
  #                 offset.text = text_spacer,
  #                 align = TRUE,
  #                 angle = "auto",
  #                 horizontal = FALSE,
  #                 hjust = 0.5,
  #                 fontsize = ggtext_size,
  #                 fontface = "bold",
  #                 family = "Helvetica"
  #                 ) +
  geom_cladelabel(
    node = 242,
    label = "I",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  geom_cladelabel(
    node = 348,
    label = "II",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  geom_cladelabel(
    node = 394,
    label = "III",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  geom_cladelabel(
    node = 363,
    label = "IV",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  geom_cladelabel(
    node = 322,
    label = "V",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  geom_cladelabel(
    node = 302,
    label = "VI",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  geom_cladelabel(
    node = 260,
    label = "VII",
    color = "black",
    offset = spacer,
    offset.text = text_spacer,
    align = TRUE,
    angle = "auto",
    horizontal = FALSE,
    hjust = 0.5,
    fontsize = ggtext_size,
    fontface = "bold",
    family = "Helvetica"
  ) +
  geom_label2(
    aes(subset = !isTip,
        label = round(as.numeric(prob), digits = 2)),
    size = 2,
    family = "Helvetica",
    label.size = NA,
    label.padding = unit(0, "mm"),
    # nudge_x = -0.1,
    # nudge_y = 0.1
  ) +
  # xlim(-0.8, NA) +
  guides(
    colour = "none",
    # fill = "none"
    fill = guide_legend()
  ) +
  theme(
    text = element_text(family = "Helvetica"),
    legend.text = element_text(face = "italic", size = 6),
    legend.title = element_blank(),
    legend.position = c(0.5, 0.47),
    # legend.position = "bottom",
    legend.text.align = 0,
    # legend.background = element_rect(fill = NA),
    legend.key.size = unit(2.5, "mm"),
    plot.margin = unit(c(0, 0, 0, 0), "mm")
  )

#log likelihood

run1 <- read_tsv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Phylogeny/2020-04-14_mrbayes/flat_prior/removed_X/no_Beagle/infile.nex.run1nB.p",
                   skip = 1) %>%
  mutate("run" = "1")

run2 <- read_tsv("/home/leonard/Dropbox/2018_PO_review/2020_rewrite/Phylogeny/2020-04-14_mrbayes/flat_prior/removed_X/no_Beagle/infile.nex.run2nB.p",
                   skip = 1) %>% 
  mutate("run" = "2")

likelihoods <- bind_rows(run1, run2) %>% 
  filter(Gen > 250000)

convergence <- ggplot(likelihoods, aes(x = Gen, y = LnL, colour = run)) +
  scale_colour_manual(values = pal_ostwald_disc[c(1,3)], name = "Run") +
  geom_line(size = 0.2) +
  labs(y = "Log-likelihood",
       x = "Generations") +
  theme_leo() +
  theme(legend.position = c(0.18, 0.925),
        legend.direction = "horizontal")

# pdf("log_likelihood.pdf", width = onecol, height = onecol * 0.5)
# convergence
# dev.off()

pdf("lac_phylogeny_supp.pdf", width = twocol * 3, height = twocol * 3)
phylo_plot + inset_element(convergence, 0.7, 0.025, 0.975, 0.1) +
  plot_annotation(tag_levels = 'A') &
  theme(plot.tag = element_text(size = 20, face = "bold", family = "Helvetica"))
dev.off()
```



### Model tanglegram

```{r}
dends <- dendlist(as.dendrogram(model_tree), as.dendrogram(clust)) %>%
  untangle(method = "step2side")

pdf("tanglegram.pdf")
tanglegram(dends,
  highlight_branches_lwd = FALSE,
  margin_inner = 5,
  margin_outer = 2
)
dev.off()
```
